contrib/win32/install/build/README.md

```diff
@@ -0,0 +1,43 @@
+# Windows OpenSSH - Build Scripts
+
+## `build-openssh.ps1`
+
+`build-openssh.ps1` is kind of a Makefile file, which first builds OpenSSH based on the passed parameters and then calls functions from the `OPenSSHBuildHelper` PowerShell module to copy to the created artifacts into a target directory or into a ZIP file.
+
+### Supported parameters
+
+| Parameter        | Required | Default                                         | Description                                                     |
+|------------------|----------|-------------------------------------------------|-----------------------------------------------------------------|
+| ProjectDirectory | yes      |                                                 | Source directory                                                |
+| Configuration    | no       | Release                                         | What configuration to build<br>Valid values: Release, Debug     |
+| Platform         | no       | x64                                             | Build for what platform<br>Valid values: x86, x64               |
+| DestinationPath  | no       | `$(OpenSSH-Bin-Path)$(Platform)\$(Configuration)` | The directory path of the location to which binaries are placed |
+| SkipClean        | no       |                                                 | Don't run a clean before a build                                |
+
+## `build-msi.ps1`
+
+`build-msi.ps1` is used for building the MSI installation package for OpenSSH. 
+
+In addition it creates MSI product GUIDs for a given version of OpenSSH in a deterministic way. Creating MSI product GUIDs in a deterministic eases the deployment of the generated MSI by using a configuration management software like Ansible, Chef or Puppet, because those frameworks idempotently configure a target system and thus rely on a fixed anchor to determine if a MSI Product of a specific version is installed or not.
+
+**Example:** Install MSI using Ansible `win_package` task
+
+```yaml
+- name: Install OpenSSH
+  win_package:
+    path: openssh.msi
+    product_id: '{0240359E-6A4C-4884-9E94-B397A02D893C}'
```

Comment:
@heaths That's the way Ansible works and I guess other Configuration Management frameworks as well (except SCCM because it's a Microsoft Product 😏😀 ). You can discuss that the people who implemented it this way did it wrong, but from my point of view it's only consequent to do it so because technically the ProductCode uniquely identifies a MSI installation package (and the product version implicitly) and as far that I understood it over the years it compares to the tuple (<packagename>,<version>) in APT or RPM. When I misunderstood this completely than it's likely that the Ansible guys did it as well and the concept of the Product Code in MSI is hard to explain and to grasp.

So, in addition I've to admit that I always wondered why WiX creates a new ProductCode (and why this should be a good thing) at every build even if the version (contents) of the product didn't change and you only needed to build a new MSI because you had to correct a spelling error in a MSI dialog for example.  In that case the product code shouldn't change because the product, you deliver using the MSI, didn't change. From my experience in the field I know that you always make great friends telling everybody your MSI has now a new product code because you changed the GUI of the MSI 😁 😁 Don't get me wrong, I definitely have less experience with MSI  than you as I can see, but the issue with those changing product codes and MSI was, and constantly is, always a pain.

That's the reason why I tried to elaborate in the README.md why I approached the subject a little different and added some code in the `build-msi.ps1` script to generate the product code based on a seed value and the version string. This procedure is standardized in chapter [4.3: Algorithm for Creating a Name-Based UUID of RFC-4122](https://tools.ietf.org/html/rfc4122#section-4.3). That way you’ll get the same product code for the same product version and the algorithm guarantees that you can calculate the product code offline by only knowing the seed value and the version. This makes the product code predicable for every party in the game who’s responsible in delivering software. I know that upgrading MSI installations is totally different beast but having the ability to calculate the product code in a foreseeable way has many advantages from my point of view (see the Ansible example).

I hope I didn't step on your toes with my statements, as we Germans are known to communicate rather directly.

Reply:
Because keeping the same ProductCode when bumping the ProductVersion constitutes a minor upgrade which you can't simply install over top of another, e.g. double click. This is all documented in the MSI docs.

If you have the product installed, you need to *reinstall* a minor upgrade, which can only be done from the command line (e.g. `msiexec /i product.msi REINSTALL=ALL REINSTALLMODE=vomus`). Obviously, that's less than ideal for customers who just want to install it. Using a chainer like WiX's burn can help simplify that story, though. That also increases size and build complexity, though, and burn was primarily meant for installing multiple projects as a single, cohesive product.

Reply:
@heaths as I tried to explain above, when you create new new version (pass a new version to the RFC-4122 algorithm) the ProductCode will do change for sure. The only difference to your favorite solution allowing WiX to a create a new product code at every build, are two facts:

1. The ProductCode is predictable. What is great from my experience. IT should be predictable, right? Randomness has definitely it's justification, but not here for my taste.
2. The ProductCode will not change when you don't change the product version (e.g. fixing a type in the UI). WiX will ALWAYS create a new PrdoctCode on every run.